### PR TITLE
Add EfficientDet

### PR DESCRIPTION
**Pull request by haossr**
**Problem**

We don't have EfficientDet model in our codebase. 

**Solution**

- Add EfficientDet implementation from https://github.com/zylo117/Yet-Another-EfficientDet-Pytorch/blob/master/backbone.py
- include model input/output mapper to make EfficietDet model compatible with Detectron format input/outputs

**Tests**

Covered by CircleCI

### CODE REVIEW

File: models/detection/__init__.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
+import numpy as np
```
Comment (jirvin16) [inline models/detection/__init__.py:1]: Is it standard to directly define all of this in `__init__.py`, rather than a separate file or set of files and importing here?

File: models/detection/efficientdet/__init__.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
+import torch
+from torch import nn
+
+from .model import BiFPN, Regressor, Classifier, EfficientNet
+from .utils import Anchors, BBoxTransform, ClipBoxes
+from .process import postprocess
+from .loss import FocalLoss
+
+
+class EfficientDetWithLoss(nn.Module):
```
Comment (jirvin16) [inline models/detection/efficientdet/__init__.py:10]: Seems like for the most modular implementation we'd want architectures and loss functions decoupled. Is there a reason to add the loss here instead (e.g. simplicity of implementation, consistency with detectron models)? I see the backbone is separate so maybe this is still possible (if so, is it possible with the detectron models too?)

File: models/detection/efficientdet/dataset.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
+import os
+import torch
+import numpy as np
+
+from torch.utils.data import Dataset, DataLoader
+from pycocotools.coco import COCO
+import cv2
+
+
+class CocoDataset(Dataset):
```
Comment (jirvin16) [inline models/detection/efficientdet/dataset.py:10]: What's the reason for including this at all? And what's the reason for including it in `models/` vs. `data/`?

File: models/detection/efficientdet/efficientnet/model.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
+import torch
```
Comment (jirvin16) [inline models/detection/efficientdet/efficientnet/model.py:1]: We already have EfficientNet models in classification.py - what's the reason for defining a new implementation here? for simplicity / to ensure compatibility with the EfficientDet implementation?

**Review by jirvin16 — APPROVED** [review]
Looks fantastic! Thanks a bunch for putting this together.

One thing I just remembered is that `ignore_width` is a really important parameter for deploying the models. EfficientDet has an implementation of this in the wind repo (as well as YOLOv3), but not sure the detectron models do. Do you have any thoughts about how to implement this?

One idea is to have all models do it the same way the wind team implemented it in EfficientDet - after the model forward pass, all predicted boxes whose center lie in the pre-specified border are dropped, and similarly all ground truth boxes with center in the border are dropped before both the loss evaluation and metric evaluation. Thoughts? Can continue this discussion in slack since not super relevant to this PR.

File: models/detection/__init__.py (hunk 1 above)
Comment (haossr) [reply models/detection/__init__.py:1]: Good point -- I think this file is getting clumsy/messy as more and more models are added. I can refactor them into small pieces. e.g. one for detectron2model and one for efficientdets.

File: models/detection/efficientdet/dataset.py (hunk 3 above)
Comment (haossr) [reply models/detection/efficientdet/dataset.py:10]: Oops, this one is not used at all; I will delete it. Good catch!

File: models/detection/efficientdet/__init__.py (hunk 2 above)
Comment (haossr) [reply models/detection/efficientdet/__init__.py:10]: Yeah, the losses are combined to be consistent with Detectron2 models (so we could use their util function and treat them in a single interface). Also, those detection models usually coupled with very specific multi-task losses and it seems to be a rare case to alter the losses -- we could still tune the coefficients between them easily.

File: models/detection/efficientdet/efficientnet/model.py (hunk 4 above)
Comment (haossr) [reply models/detection/efficientdet/efficientnet/model.py:1]: This EfficientDet implementation coupled with its own efficientnet version. 
The author argued other implementations have flaws that prevent him/her from achieving a similar performance as reported. 

Actually I am thinking of retiring/altering the efficientnet version in the repo since this one seems to be more memory efficient. We could open an issue for that.

File: models/detection/efficientdet/__init__.py (hunk 2 above)
Comment (jirvin16) [reply models/detection/efficientdet/__init__.py]: Makes sense - sgtm!

File: models/detection/efficientdet/efficientnet/model.py (hunk 4 above)
Comment (jirvin16) [reply models/detection/efficientdet/efficientnet/model.py:1]: Got it! Also sounds good.
